base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+        TransactionProtocolState::SenderSetup(self)
+    }
+
+    fn set_nonce(mut self, nonce: SecretKey) -> TransactionProtocolState {
+        self.state_data.sender_private_nonce = Some(nonce.clone());
+        self.state_data.sender_public_nonce = Some(PublicKey::from_secret_key(&nonce));
+        TransactionProtocolState::SenderSetup(self)
+    }
+
+    /// This is the final call you make in this state that will transition you to the next state
+    fn set_offset(mut self, offset: BlindingFactor) -> Result<TransactionProtocolState, TransactionProtocolError> {
+        // Validate the current state to check we can proceed to the next state
+        if self.state_data.amount.is_none() ||
+            self.state_data.lock_height.is_none() ||
+            self.state_data.fee.is_none() ||
+            self.state_data.offset.is_some() ||
+            self.state_data.inputs.len() == 0 ||
+            self.state_data.sender_private_nonce.is_none()
+        {
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        // Validate that inputs, outputs, fees and amount balance
+        let mut sum = &CommitmentFactory::create(&SecretKey::default(), &SecretKey::from(self.state_data.fee.unwrap())) +
+            &CommitmentFactory::create(&SecretKey::default(), &SecretKey::from(self.state_data.amount.unwrap()));
+        for o in self.state_data.outputs.clone() {
+            sum = &sum + &o.commitment;
+        }
+        for i in self.state_data.inputs.clone() {
+            sum = &sum - &i.commitment;
+        }
+        sum = &sum -
+            &CommitmentFactory::create(
+                &self.state_data.sender_excess_blinding_factor.unwrap().into(),
+                &SecretKey::default(),
+            );
+
+        if sum != CommitmentFactory::create(&SecretKey::default(), &SecretKey::default()) {
+            return Err(TransactionProtocolError::ValidationError);
+        }
+
+        // If validation passes, select offset and move on to next state
+        self.state_data.sender_excess_blinding_factor =
+            Some(&self.state_data.sender_excess_blinding_factor.unwrap() - offset);
+        self.state_data.offset = Some(offset);
+        self.state_data.sender_excess =
+            Some(PublicKey::from_secret_key(&self.state_data.sender_excess_blinding_factor.unwrap()));
+
+        Ok(TransactionProtocolState::ReceiverOutputSetup(ReceiverOutputSetup::new(self.state_data)))
+    }
+}
+
+/// In this state the Receiver will provide the data for their receiving output which will transition them to the next
+/// state.
+/// In this state the Sender can construct their public data message to be sent to the Receiver AND the Sender will wait
+/// in this state to accept the public data from the Receiver to advance to the next state.
+#[derive(Clone, Debug)]
+struct ReceiverOutputSetup {
+    state_data: TransactionProtocolStateData,
```

Comment:
Box?

Reply:
See comment above (Vectors are already on the heap)